src/csharp/Grpc.Core/Internal/BatchContextSafeHandle.cs

```diff
@@ -101,6 +118,24 @@ public ClientSideStatus GetReceivedStatusOnClient()
             return new ClientSideStatus(status, metadata);
         }
 
+        // Gets data of recv_status_on_client completion without throwing an
+        // exception if the data is corrupt. This is to stop UnaryCall from hanging
+        // if there is some internal problem.
+        // See https://github.com/grpc/grpc/issues/29854
+        public ClientSideStatus TryGetReceivedStatusOnClient()
```

Comment:
Instead of introducing a new method, we should modify the original GetReceivedStatusOnClient().

AFAIK The only problem experienced in #29854 was the inability to convert the detailsPtr to details string (via 
MarshalUtils.PtrToStringUTF8)

I think we should simply have a  try - catch around `MarshalUtils.PtrToStringUTF8(detailsPtr, (int)detailsLength.ToUInt32());` (and possibly also around `string debugErrorString = Marshal.PtrToStringAnsi(...)`) and log an error if the conversion fails (which should never happen under normal circumstances).

https://github.com/grpc/grpc/blob/90b0ea387e988b8615882e152a5152ed0801ec06/src/csharp/Grpc.Core/Internal/BatchContextSafeHandle.cs#L94

Reply:
One way you could do this is to wrap the MarshalUtils.PtrToStringUTF8 into a new private method (e.g. `PtrToStringUTF8WithSafeFallback` and call it instead of MarshalUtils.PtrToStringUTF8 here.
https://github.com/grpc/grpc/blob/90b0ea387e988b8615882e152a5152ed0801ec06/src/csharp/Grpc.Core/Internal/MarshalUtils.cs#L37